core/src/main/scala/org/apache/spark/rpc/netty/Dispatcher.scala

```diff
@@ -0,0 +1,226 @@
+      if (inbox != null) {
+        val rpcCallContext =
+          new LocalNettyRpcCallContext(
+            inbox.endpointRef, message.senderAddress, message.needReply, p)
+        postMessageToInbox(inbox,
+          ContentMessage(message.senderAddress, message.content, message.needReply, rpcCallContext))
+        return
+      }
+    }
+    p.tryFailure(
+      new SparkException(s"Could not find ${message.receiver.name} or it has been stopped"))
+  }
+
+  private def postMessageToInbox(inbox: Inbox, message: InboxMessage): Unit = synchronized {
+    if (stopped) {
+      logWarning(s"Drop ${message} because RpcEnv has been stopped")
+      return
+    }
+    inbox.post(message)
+    receivers.put(inbox.endpoint)
+  }
+
+  private[netty] class MessageLoop extends Runnable {
+    override def run(): Unit = {
+      try {
+        while (true) {
+          try {
+            val endpoint = receivers.take()
+            if (endpoint == DummyEndpoint) {
+              // Put DummyEndpoint back so that other MessageLoops can see it.
+              receivers.put(DummyEndpoint)
+              return
+            }
+            val inbox = endpointToInbox.get(endpoint)
+            if (inbox != null) {
+              val inboxStopped = inbox.process(Dispatcher.this)
+              if (inboxStopped) {
+                endpointToInbox.remove(endpoint)
+              }
+            } else {
+              // The endpoint has been stopped
+            }
+          } catch {
+            case NonFatal(e) => logError(e.getMessage, e)
+          }
+        }
+      } catch {
+        case ie: InterruptedException => // exit
+      }
+    }
+  }
+
+  private val parallelism = nettyEnv.conf.getInt("spark.rpc.netty.dispatcher.parallelism",
+    Runtime.getRuntime.availableProcessors())
+
+  private val executor = ThreadUtils.newDaemonFixedThreadPool(parallelism, "dispatcher-event-loop")
+
+  (0 until parallelism) foreach { _ =>
+    executor.execute(new MessageLoop)
+  }
+
+  def stop(): Unit = {
+    synchronized {
+      if (stopped) {
+        return
+      }
+      stopped = true
+    }
+    // When we reach here, other threads won't update `nameToEndpoint`. So we can guarantee all
+    // registered endpoints will be stopped correctly.
+    for (name <- nameToEndpoint.keySet().asScala) {
+      unregisterRpcEndpoint(name)
+    }
+    // When we reach here, the new items put into receivers will always be DummyEndpoint, others
+    // will be rejected. So that we can make sure we will process all messages that have already in
+    // the Inboxes.
+    receivers.put(DummyEndpoint)
+    executor.shutdown()
+  }
+
+  def awaitTermination(): Unit = {
+    executor.awaitTermination(Long.MaxValue, TimeUnit.MILLISECONDS)
+  }
+
+  /**
+   * Return if the endpoint exists
+   */
+  def verify(name: String): Boolean = {
+    nameToEndpoint.containsKey(name)
+  }
+
+}
+
+/**
+ * A dummy endpoint that indicates MessageLoop should exit its loop.
```

Comment:
This is generally called a "poison", not a "dummy". Also, can this be a private class inside `Dispatcher` instead?